designs/mixins.md

```diff
@@ -0,0 +1,1075 @@
+@twoTrait
+structure Two {}
+
+/// Three
+@mixin
+@foo(3)
+@threeTrait
+structure Three with One, Two {}
+
+/// Four
+@fourTrait
+structure FourStruct with Three {}
+```
+
+Is equivalent to the following flattened structure:
+
+```
+/// Four
+@oneTrait
+@twoTrait
+@threeTrait
+@fourTrait
+@foo(3)
+structure FourStruct with Three {}
+```
+
+#### Mixin local traits
+
+Sometimes it's useful to apply traits to a mixin that are not copied onto
+shapes that use the mixin. For example, if a mixin is an implementation detail
+of a model, then it is recommended to apply the `@private` trait to the mixin
+so that shapes outside of the namespace the mixin is defined within cannot
+refer to the mixin. However, every shape that uses the mixin doesn't
+necessarily need to be marked as `@private`. The `localTraits` property of
+the `@mixin` trait can be used to ensure that a list of traits applied to
+the mixin are not copied onto shapes that use the mixin (note that this has
+no effect on the traits applied to members contained within a mixin).
+
+Consider the following model:
+
+```
+namespace smithy.example
+
+@private
+@mixin(localTraits: [private])
+structure PrivateMixin {
+    foo: String
+}
+
+structure PublicShape with PrivateMixin {}
+```
+
+The `PrivateMixin` shape can only be referenced from the `smithy.example`
+namespace. Because `private` is present in the `localTraits` property of the
+`@mixin` trait, `PublicShape` is not marked with the `@private` trait and
+can be referred to outside of `smithy.example`. `PublicShape` is equivalent
+to the following flattened structure:
+
+```
+structure PublicShape {
+    foo: String
+}
+```
+
+> See *Trait definitions* for a full description of `localTraits`.
+
+
+### Overriding traits on mixin members
+
+The members and traits applied to members of a mixin are copied onto the target
+shape. It is sometimes necessary to provide a more specific trait value for a
+copied member or to add traits only to a specific copy of a member. This can be
+achieved in one of two ways: by redefining a member or using `apply`
+statements.
+
+
+#### Redefining members
+
+1. The redefined member MUST have the exact same name as the mixin member.
+2. The member MUST target the exact same shape as the mixin member.
+3. The `@override` trait MUST be applied to any member that redefines a mixin
+   member. It is an error if a member marked with the `@override` trait does
+   not actually redefine a mixin member copied into a shape.
+4. Just like overriding traits on the containing shape, any trait applied
+   to a redefined member completely overrides any resolved traits applied to
+   the mixin member.
+
+For example, the definition of the `ABC` shape in the following model:
+
+```
+@mixin
+structure A {
+    /// A docs
+    @required
+    a: String
+}
+
+@mixin
+structure B with A {
+    /// B docs
+    @tags(["hello"])
+    @deprecated
+    b: String
+}
+
+structure ABC with A, B {
+    /// More specific B docs
+    @override
```

Comment:
In this example we alter an existing trait, I can see how one would add one, but is it possible to remove a trait?

Reply:
There is not. We could allow traits to be filtered per member by extending the `@override` trait. However, I think that needing to _remove_ a trait from a specific member implies a case where the mixin is not appropriate, and the site of the mixin isn't actually opting into the mixin. Adding this kind of filtering per member, I think, would make the models harder to understand (plus harder to implement in Smithy implementations).

An alternative to removing traits but still leveraging mixins is to either not use a mixin for the one-off cases, or to make two layers of mixins. For example, say that some applications of a mixin want a member to be required, and others to be optional. In that case, make two mixins:

```
@mixin
structure FooMixinOptional {
    someMember: String
    otherMember: String
}

@mixin
structure FooMixinRequired with FooMixin {
    @override
    @required
    someMember: String
}
```